### ISSUE TITLE
clear tube command

### ISSUE THREAD

**Issue by kr** [beanstalkd/beanstalkd#25]
http://groups.google.com/group/beanstalk-talk/t/4ed1d368d7b3a5a8

**Comment by geudrik** [comment]
Yea, flushing a tube would be nice. The whole queue while we're at it.

**Comment by prgTW** [comment]
:+1: +1 to this idea

**Comment by JensRantil** [comment]
@pentium10 Interesting it takes such a long time for you to simply delete your jobs. My little test script does that pretty quickly:

```python
import sys
import beanstalkc

beanstalk = beanstalkc.Connection(host='localhost', port=11300)

if sys.argv[1] == "fill":
  print "Filling"
  for i in range(711769):
    beanstalk.put("hello")
elif sys.argv[1] == "clear":
  run = True
  while run:
    job = beanstalk.reserve(timeout=0)
    if job is None:
      run = False
    else:
      job.delete()
```

Micro-benchmark:
```bash
$ time ./testpy/bin/python test.py fill                                                                                             [20:01:23]
ERROR:root:Failed to load PyYAML, will not parse YAML
Filling
./testpy/bin/python test.py fill  18.73s user 11.42s system 57% cpu 52.188 total
$ time ./testpy/bin/python test.py clear                                                                                            [20:02:06]
ERROR:root:Failed to load PyYAML, will not parse YAML
./testpy/bin/python test.py clear  41.44s user 22.28s system 57% cpu 1:49.88 total
```

Any idea why clearing the tube is so slow for you?

**Comment by pentium10** [comment]
My jobs are around 800kbytes, maybe try with that size.

**Comment by JensRantil** [comment]
@pentium10 711769*800 kbytes = ~543 GBytes. Just making sure here, are you sure you had that many items in memory? Pretty beefy server, huh? ;) Also, are you sure you weren't swapping memory to disk?

**Comment by pentium10** [comment]
I will double check, we have only 64GB.

**Comment by terion-name** [comment]
absolutely needed feature

**Comment by JensRantil** [comment]
Could everyone asking for this feature explain why you guys simply can't do what my script in https://github.com/kr/beanstalkd/issues/25#issuecomment-205023390 does. That is, simply reserve and delete all jobs one by one. Beanstalkd is very fast.

**Comment by prgTW** [comment]
@JensRantil really? I could think at least few actually:
- atomicity of operation
- minimizing Round-Trip-Time (see [pipelining feature of Redis](https://redis.io/topics/pipelining))
- huge speed improvement of the operation itself due to possible optimization of the process of clearing a tube instead of deleting items one-by-one
- better tolerance for errors like net downtime, packets being lost etc.

It's easy to say that beanstalk is fast having it installed on localhost and doing benchmarks on it but c'mon, not all of us install beanstalk on the same machine as the executor scripts. Imagine having 20k queued messages and RTT time even as low as only 1ms. You're wasting 20 seconds just for RTT! Now think of the same situation when RTT time is 10ms (I abstract the reasons of this situation). That's 200 seconds just for RTT time. Think of the probability of error doing 20k requests instead of just 1.

**Comment by JensRantil** [comment]
@prgTW Thanks for taking the time to answer. Please consider your tone. Stuff like "really?" and "c'mon" isn't gonna help your point through better.

I agree with you that the operation clearly would be faster than emptying the queue one by one. That said, one benefit of emptying the queue one-by-one is that it follow the same process as general consumption which will allow other consumers to not be surprised that their current task was deleted by someone else.

Regarding your RTT time you are assuming your client will execute a single operation at a time, but you should be able to do pipelining with beanstalkd just like you do in Redis. It should give you like a 100x speedup depending on your batch size of operations. https://redis.io/topics/pipelining might be a nice reference if you want to read up on it.

> Think of the probability of error doing 20k requests instead of just 1.

Yeah, but that's what we have TCP for, right? :-)

**Comment by prgTW** [comment]
> Yeah, but that's what we have TCP for, right? :-)

Consider also web server timeouts, application timeouts, keepalive timeouts or any other system/application error that might come in unexpectedly during this long running process. I personally would appreciate the atomicity benefit from "clear tube" command.

Sorry for the tone but showing benchmarks on localhost (an idyllic environment) with every message as big as few bytes was a bit naive, but either way I apologize :)

**Comment by ysmolsky** [comment]
This feature looks good to me. From performance point of view, it's a must.

**Comment by HorseLuke** [comment]
I'd like to have this feature too.
Now I am using reserve(0) + delete(job_id) to simulate clear tube, but it can not delete buried or not ready job.

**Comment by SamMousa** [comment]
Hmm, couldn't this be implemented in some kind of transparent fashion?

- Tubes are stored in a list
- A linear search on this list is done using `tube_find`

Implementation:
1. Create new tube with the same name, prepend it to the list of tubes -- this will make new connections use the new empty tube.
2. Iterate over connections update `->use` to use the new tube.
3. Iterate over connections update `->watch` to use the new tube if they contain the old tube.

At this point we're essentially done, except for the fact that the old tube still lingers around, jobs in it can probably still be reserved using `reserve-job`. So what we do next:
1. Delete all jobs in the old tube that are not reserved
2. Disable new reservations on the tube
3. Run some cleanup code that once in a while checks the deleted tube(s) for remaining reserved jobs, as soon as it reaches 0 delete the tube.

This would be totally transparent to client consuming messages and to producers producing messages.

**Comment by kr** [comment]
Yeah @SamMousa, your plan [swap out the fixed-size tube structure and incrementally delete the jobs later] would probably work. I would do something along those lines, except I would not make a new tube struct.

Instead I would move the job array/linked list pointers out of the tube's `ready` & `delay` & `buried` fields, into a "to be deleted" data structure somewhere. The idea is to reset the two `Heap` objects and the linked list so they're empty, and leave the existing tube struct where it is. That would make it easier to preserve invariants (eg there should be exactly one tube in the list with the given name as long as its refcount is positive, and zero otherwise).

You would also need to add a record to the binlog telling the new process to clear the tube.

I think there is still an open question whether the clear tube command should let any of the tube's jobs stick around depending on what state they're in. (buried? delayed?) Or should there be a parameter to control this behavior? I don't have a strong intuition for this aspect, but it seems that most people so far are assuming that all jobs in the tube, regardless of their state, should be deleted. So maybe that is the best thing to do by dint of it being what everyone expects.

**Comment by SamMousa** [comment]
I don't actually speak C so the things you're considering sound good to me ;-)
